pkg/driver/controller_server.go

```diff
@@ -1350,28 +1369,43 @@ func (driver *Driver) ListSnapshots(ctx context.Context, request *csi.ListSnapsh
 	// request.StartingToken <- use me
 	// request.MaxEntries <- use me
 
-	// If driver doesn't know any storage providers yet, then return empty list with success
-	if len(driver.storageProviders) == 0 {
-		log.Info("No storage providers are known to the CSI driver yet.")
-		return &csi.ListSnapshotsResponse{
-			Entries: []*csi.ListSnapshotsResponse_Entry{},
-		}, nil
-	}
-
-	// TODO: ListSnapshots does not provide Secrets nor a NodeId so it simply needs to return information about all of the
-	// snapshots it knows about.  This can be implemented by reading all snapshots from every array known to the driver.  We can do this
-	// at initialization time and keep them in memory.
 	var allSnapshots []*model.Snapshot
-	for _, storageProvider := range driver.storageProviders {
-		// TODO: skip storage providers that are not of the same vendor... or not.  Need to think through this
-		snapshots, err := storageProvider.GetSnapshots(request.SourceVolumeId)
+
+	// Secrets are optional
+	if request.Secrets != nil {
+		storageProvider, err := driver.GetStorageProvider(request.Secrets)
 		if err != nil {
-			log.Trace("err: ", err.Error())
-			return nil, status.Error(codes.Internal, "Error while attempting to list snapshots")
+			log.Error("err: ", err.Error())
+			return nil, status.Error(codes.Unavailable, fmt.Sprintf("Failed to get storage provider from secrets, %s", err.Error()))
 		}
-		log.Tracef("Read %d snapshots from a storage provider", len(snapshots))
-		for _, snapshot := range snapshots {
-			allSnapshots = append(allSnapshots, snapshot)
+		allSnapshots, err = getSnapshotsForStorageProvider(ctx, request, storageProvider)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// In the event that ListSnapshots does not provide Secrets it simply needs to return information about all of the
+		// snapshots it knows about.  This can be implemented by reading all snapshots from every array known to the driver.  We can do this
+		// at initialization time and keep them in memory.
+
+		// If driver doesn't know any storage providers yet, then return empty list with success
+		if len(driver.storageProviders) == 0 {
+			log.Info("No storage providers are known to the CSI driver yet.")
+			return &csi.ListSnapshotsResponse{
+				Entries: []*csi.ListSnapshotsResponse_Entry{},
+			}, nil
+		}
+
+		for _, storageProvider := range driver.storageProviders {
+			snapshots, err := getSnapshotsForStorageProvider(ctx, request, storageProvider)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(snapshots) > 0 {
```

Comment:
Since it seems like we always have either snapshotID or sourceVolumeID, can we break the loop here once we get snapshots from one provider? If both of these can be nil, then do we still need to fetch all snapshots of all volumes and all providers in that case?

Reply:
We sure can.  Thanks for that idea.